### PR TITLE
DRY up circleci config

### PR DESCRIPTION
**Pull request by bengl**
### What does this PR do?
Makes heavy use of reverence variables, starts using matrices, adds a helpful command.

### Motivation

The less things to edit, the less that can be broken when we add new things. We can also eventually use the matrices approach for testing all the pugins across versions of node.

### Additional Notes

See the individual commit messages for details on what changes.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -53,41 +62,42 @@ node-core-base: &node-core-base
...
   working_directory: ~/dd-trace-js
   steps:
-    - checkout
-    - *yarn-versions
-    - *restore-yarn-cache
-    - *yarn-install
-    - *save-yarn-cache
+    - checkout-and-yarn-install
+    - run:
+        name: Get plugin name
```
Comment (rochdev) [inline .circleci/config.yml:95]: This is already done in one of the script. It would be more consistent to do it always in the scripts or always in the CircleCI config, not both.

File: .circleci/config.yml [hunk 2]
```diff
@@ -172,18 +172,17 @@ node-plugin-base: &node-plugin-base
...
   steps:
-    - checkout
-    - *yarn-versions
-    - *restore-yarn-cache
-    - run:
-        name: Install dependencies
-        command: yarn install
-    - *save-yarn-cache
+    - checkout-and-yarn-install:
+        yarn-args: ""
```
Comment (rochdev) [inline .circleci/config.yml:185]: Why is this needed?

File: .circleci/config.yml [hunk 3]
```diff
@@ -1265,71 +922,70 @@ workflows:
...
+      - node-tedious: *matrix-supported-node-versions
+      - node-when: *matrix-supported-node-versions
+      - node-winston: *matrix-supported-node-versions
       - codecov:
           requires:
-            - node-core-12
-            - node-core-14
-            - node-core-16
-            - node-core-latest
+            - node-core
```
Comment (rochdev) [inline .circleci/config.yml:988]: Does our current `codecov` usage work with matrixes?

File: .circleci/config.yml [hunk 4]
```diff
@@ -1513,66 +1151,57 @@ workflows:
               only:
                 - master
     jobs:
-      - node-core-12
-      - node-core-14
-      - node-core-16
-      - node-core-latest
+      - node-core: *matrix-supported-node-versions
```
Comment (rochdev) [inline .circleci/config.yml:1154]: Can this list be extracted to a YAML anchor so that it doesn't need to be repeated between `build` and `nightly`?

Comment (rochdev) [reply .circleci/config.yml:1154]: Basically like the benchmark jobs.

**Review by rochdev — CHANGES_REQUESTED** [review]
A few small things but otherwise LGTM.

File: scripts/install_plugin_modules.js [hunk 5]
```diff
@@ -25,6 +25,9 @@ function assertVersions () {
   let filter = []
   let names = Object.keys(plugins)
 
+  if (process.env.PLUGINS && process.env.PLUGINS.trim().length === 0) {
+    delete process.env.PLUGINS
```
[the remark this answers is not in the document]
Comment (bengl) [reply scripts/install_plugin_modules.js:29]: It seemed to be the easiest approach in conjunction with this https://github.com/DataDog/dd-trace-js/commit/3c83f6e4cde676e2ae18ca860c3d03b456338131#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R96

The idea was to simplify sirun tests which currently all have `PLUGINS` equal to what their name is if it's a plugin test, but some aren't plugin tests so the value ends up being an empty string.

File: .circleci/config.yml (hunk 1 above)
Comment (bengl) [reply .circleci/config.yml:95]: Which script? AFAICT, no script is grabbing the plugin name from the `CIRCLE_JOB` env var.

File: .circleci/config.yml (hunk 2 above)
Comment (bengl) [reply .circleci/config.yml:185]: The base case includes `--ignore-scripts`. That's why it typically wasn't reused, and instead we'd end up with stuff like this all over the place: https://github.com/DataDog/dd-trace-js/pull/1505/files/a5eb8c09fed6c7fc466160819eff96958e9d7396#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47L185

File: .circleci/config.yml (hunk 4 above)
Comment (bengl) [reply .circleci/config.yml:1154]: Currently, the nightly doesn't include coverage reports. YAML anchors can't be used to extend arrays. If you're fine with just running coverage nightly, then we can re-use it. What do you think?

Comment (rochdev) [reply .circleci/config.yml:1154]: I think for now it's fine if it makes this simpler. We can always reorganize later with more workflows if needed.

File: .circleci/config.yml (hunk 2 above)
Comment (rochdev) [reply .circleci/config.yml:185]: That link doesn't work for me, but what's the problem with just keeping the `--ignore-scripts` everywhere?

File: .circleci/config.yml (hunk 1 above)
Comment (rochdev) [reply .circleci/config.yml:95]: https://github.com/DataDog/dd-trace-js/pull/1505/files#diff-2e17a4cf3758f4f2c6dc40393c8d6bef4c5d77c1b67a342c2046239c8dc02ea9

File: scripts/install_plugin_modules.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (rochdev) [reply scripts/install_plugin_modules.js:29]: But why is it needed to delete the environment variable?

[the remark this answers is not in the document]
Comment (bengl) [reply scripts/install_plugin_modules.js:29]: It prevents a TypeError on line 32. An alternative would be just to change the conditional on line 31. Would that be preferable?

File: .circleci/config.yml (hunk 2 above)
Comment (bengl) [reply .circleci/config.yml:185]: 🤷‍♂️ I dunno. We didn't previously have it everywhere. instead we put manual `yarn install` commands in everywhere. I just kept the same effect with this.

File: scripts/install_plugin_modules.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (rochdev) [reply scripts/install_plugin_modules.js:29]: I'd say yes, it makes the intent more obvious and simplifies the conditions.

File: .circleci/config.yml (hunk 2 above)
Comment (rochdev) [reply .circleci/config.yml:185]: Ok, it's fine to leave it like that then, but I think it wasn't on purpose that there was multiple ways to call `yarn install`.

File: scripts/install_plugin_modules.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bengl) [reply scripts/install_plugin_modules.js]: done

File: .circleci/config.yml (hunk 4 above)
Comment (bengl) [reply .circleci/config.yml]: done

File: .circleci/config.yml (hunk 3 above)
Comment (bengl) [reply .circleci/config.yml:988]: I don't know if everything is in there, but coverage hasn't gone *down* on this PR if I'm reading it right, so if we want we can give it some further effort in a future PR. How does that sound?

**Review by rochdev — APPROVED** [review]

**Review by rochdev — DISMISSED** [review]

**Review by rochdev — APPROVED** [review]
